### PR TITLE
macos: implement async getaddrinfo and getnameinfo

### LINKED ISSUE #2349: darwin: use getaddrinfo_async_start()
libSystem (actually libinfo) provides an asynchronous getaddrinfo(). There is no public header that I can find but the prototypes are as follows (from libinfo's netdb_async.h):
```c
/*
 * getaddrinfo
 */
typedef void (*getaddrinfo_async_callback)(int32_t status, struct addrinfo *res, void *context);
int32_t getaddrinfo_async_start(mach_port_t *p, const char *nodename, const char *servname, const struct addrinfo *hints, getaddrinfo_async_callback callback, void *context);
int32_t getaddrinfo_async_send(mach_port_t *p, const char *nodename, const char *servname, const struct addrinfo *hints);
int32_t getaddrinfo_async_receive(mach_port_t p, struct addrinfo **res);
int32_t getaddrinfo_async_handle_reply(void *msg);
void getaddrinfo_async_cancel(mach_port_t p);

 
/*
 * getnameinfo
 */
typedef void (*getnameinfo_async_callback)(int32_t status, char *host, char *serv, void *context);
int32_t getnameinfo_async_start(mach_port_t *p, const struct sockaddr *sa, size_t salen, int flags, getnameinfo_async_callback callback, void *context);
int32_t getnameinfo_async_send(mach_port_t *p, const struct sockaddr *sa, size_t salen, int flags);
#if (!defined(LIBINFO_INSTALL_API) || !LIBINFO_INSTALL_API)
int32_t getnameinfo_async_receive(mach_port_t p, char **host, char **serv);
#endif
int32_t getnameinfo_async_handle_reply(void *msg);
void getnameinfo_async_cancel(mach_port_t p);
```
The API works by returning a `mach_port_t` (which is just an `unsigned int`) in the `p` out param that can then be polled with `EVFILT_MACHPORT`.

It's supported by libinfo 392.1, which is shipped with OS X 10.7, the oldest version libuv supports (but I think it existed even before that.)

If need be, libuv could dlsym() the symbols and fall back to the thread pool if they're missing - but I don't think that's necessary.

Nice-to-have low prio that would free up the thread pool for more useful work.

### PR DESCRIPTION
**Pull request by cabelitos**
This commit allows OS X/iPhone users to enable these
two function using non-blocking calls, which allows
libuv to save resources by not creating threads to
handle these operations.

Closes #2349

### CODE REVIEW

File: test/test-threadpool-cancel.c [hunk 1]
```diff
@@ -184,12 +184,19 @@ TEST_IMPL(threadpool_cancel_getaddrinfo) {
...
 
+#if defined(__APPLE__) && defined(_USE_ASYNC_ADDR_INFO)
+  # define TIMEOUT 0
+#else
+  # define TIMEOUT 10
+#endif
+
   ASSERT(0 == uv_timer_init(loop, &ci.timer_handle));
-  ASSERT(0 == uv_timer_start(&ci.timer_handle, timer_cb, 10, 0));
+  ASSERT(0 == uv_timer_start(&ci.timer_handle, timer_cb, TIMEOUT, 0));
```
Comment (cabelitos) [inline test/test-threadpool-cancel.c:194]: In case async addr is enable, the "task" is not considered as low prio anymore. Therefore the timer might be scheduled after uv_getaddrinfo completes. In order to properly cancel the task a timer with timeout 0 is created.

File: configure.ac [hunk 2]
```diff
@@ -54,6 +54,10 @@ AC_SEARCH_LIBS([clock_gettime], [rt])
 AC_SEARCH_LIBS([sendfile], [sendfile])
 AC_SEARCH_LIBS([socket], [socket])
 AC_SYS_LARGEFILE
+AC_ARG_ENABLE(use-osx-async-libinfo,
+              AS_HELP_STRING([--disable_use_osx_async_libinfo],
```
Comment (cabelitos) [inline configure.ac:58]: Should be - and not _

File: CMakeLists.txt [hunk 3]
```diff
@@ -32,6 +32,11 @@ endif()
 
 option(ASAN "Enable AddressSanitizer (ASan)" OFF)
 option(TSAN "Enable ThreadSanitizer (TSan)" OFF)
+option(USE_OSX_ASYNC_LIBINFO "Enable async getnameinfo on OSX" ON)
+if(USE_OSX_ASYNC_LIBINFO AND APPLE)
```
Comment (cabelitos) [inline CMakeLists.txt:36]: missing check for libinfo itself.

File: src/unix/getnameinfo.c [hunk 4]
```diff
@@ -27,6 +27,49 @@
...
+extern int32_t (getnameinfo_async_start)(mach_port_t *p, const struct sockaddr *, size_t, int, getnameinfo_async_callback, void *);
+extern int32_t (getnameinfo_async_handle_reply)(void *);
+extern void (getnameinfo_async_cancel)(mach_port_t);
+
+static void uv__getnameinfo_async_done(int32_t status, char *host, char *service, void *context) {
+  uv_getnameinfo_t *req;
+
+  req = context;
+  if (host)
+    uv__strscpy(req->host, host, sizeof(host));
```
Comment (cabelitos) [inline src/unix/getnameinfo.c:44]: sizeof(req->host)

File: src/unix/getnameinfo.c [hunk 5]
```diff
@@ -27,6 +27,49 @@
...
+  if (service)
+    uv__strscpy(req->service, service, sizeof(service));
```
Comment (cabelitos) [inline src/unix/getnameinfo.c:46]: sizeof(req->service)

**Comment by cabelitos** [comment]
Left some notes to myself while I wait for the maintainers review. Once the code is reviewed I'll send the fixes.

Thank you.

File: test/test-threadpool-cancel.c (hunk 1 above)
Comment (bnoordhuis) [reply test/test-threadpool-cancel.c:194]: I think I follow your line of reasoning. This test saturates the thread pool but the request is not going through the thread pool anymore.

I'm okay with changing the timeout to 0 unconditionally, no compile-time guards, as long as you add a comment explaining why it's 0.

File: CMakeLists.txt (hunk 3 above)
Comment (barbieri) [reply CMakeLists.txt:36]: @bnoordhuis AFAIU linkers do lazy-load by default these days, I did check and https://linux.die.net/man/1/ld says `-z lazy` is the default, it's only needed if you `-z now` to force it to resolve symbols immediately.

Going with `dlopen` is not that hard, but adds a bulk of code that could be avoided by using the correct flags.

Comment (bnoordhuis) [reply CMakeLists.txt:36]: The reason we do that is to make life easier to use for projects linking statically against libuv. No need for the downstream project to link against libinfo, or CoreFoundation, or IOKit, or...

File: src/unix/getnameinfo.c [hunk 6]
```diff
@@ -107,12 +158,31 @@ int uv_getnameinfo(uv_loop_t* loop,
...
+
+    get_nameinfo_result = getnameinfo_async_start(&port,
+                                                  (struct sockaddr*) &req->storage,
+                                                  salen,
+                                                  req->flags,
+                                                  uv__getnameinfo_async_done,
+                                                  req);
+    if (get_nameinfo_result == -1) {
+      return get_nameinfo_result;
+    }
```
[the remark this answers is not in the document]
Comment (barbieri) [reply src/unix/getnameinfo.c:173]: it makes no sense the thread pool to be used, if one fails the other will fail as well. Other than that, ok.

File: CMakeLists.txt (hunk 3 above)
Comment (barbieri) [reply CMakeLists.txt:36]: but I think detecting the library with https://cmake.org/cmake/help/latest/command/find_library.html is the way to go to avoid misusage

File: src/unix/getaddrinfo.c [hunk 7]
```diff
@@ -35,6 +35,44 @@
...
 #include <netdb.h>
 
+static void uv__getaddrinfo_common_done(uv_getaddrinfo_t* req, int status);
+
+#if defined(_USE_LIBINFO)
+
+typedef void (getaddrinfo_async_callback)(int32_t, struct addrinfo *, void *);
+extern int32_t (getaddrinfo_async_start)(mach_port_t *, const char *, const char *, const struct addrinfo *, getaddrinfo_async_callback, void *);
+extern int32_t (getaddrinfo_async_handle_reply)(void *);
+extern void (getaddrinfo_async_cancel)(mach_port_t);
```
[the remark this answers is not in the document]
Comment (barbieri) [reply src/unix/getaddrinfo.c:45]: @cabelitos if you remove the extern, at least leave a comment saying they come from a header that is not installed by default (link to it?)

File: include/uv/darwin.h [hunk 8]
```diff
@@ -30,9 +30,25 @@
...
 #endif
 
+#if defined(_USE_LIBINFO)
+# define UV_IO_PRIVATE_PLATFORM_EXTRA_FIELDS                                  \
+  void *data;
+
+# define UV_GETADDRINFO_PLATFORM_PRIVATE_EXTRA_FIELDS                         \
+    uv__io_t async_addrinfo_io;
+# define UV_GETNAMEINFO_PLATFORM_PRIVATE_EXTRA_FIELDS                         \
+    uv__io_t async_getnameinfo_io;
```
[the remark this answers is not in the document]
Comment (barbieri) [reply include/uv/darwin.h:40]: @bnoordhuis here you commented on the extension of `uv_getnameinfo_t/uv_getaddrinfo_t`, but I guess you meant the one on top `UV_IO_PRIVATE_PLATFORM_EXTRA_FIELDS` which is a `void *data`, am I correct?

File: src/unix/getaddrinfo.c [hunk 9]
```diff
@@ -203,12 +244,30 @@ int uv_getaddrinfo(uv_loop_t* loop,
...
+                                                  req->hostname,
+                                                  req->service,
+                                                  req->hints,
+                                                  uv__getaddrinfo_async_done,
+                                                  req);
+    if (get_addrinfo_result == -1) {
+      return get_addrinfo_result;
+    }
+    uv__io_init(&req->async_addrinfo_io, uv__getaddrinfo_async_work, port);
+    req->async_addrinfo_io.data = req;
```
Comment (barbieri) [inline src/unix/getaddrinfo.c:260]: @cabelitos do we really need this? I saw @bnoordhuis comment about `data`, but since `req->async_addrinfo_io` is contained in `req`, to obtain the `req` given `async_addrinfo_io` it's a simpler pointer arithmetic as done by `container_of()`.

Comment (barbieri) [reply src/unix/getaddrinfo.c:260]: IOW: you don't do anything in here, when you need to use from `uv__getaddrinfo_async_work` you do:

```c
uv_getaddrinfo_t* req = container_of(w, uv_getaddrinfo_t, async_addrinfo_io);
```

File: src/unix/getaddrinfo.c [hunk 10]
```diff
@@ -35,6 +35,44 @@
...
+    req->retcode = status;
+  else
+    req->retcode = uv__getaddrinfo_translate_error(status);
+  uv__io_close(req->loop, &req->async_addrinfo_io);
+  uv__req_unregister(req->loop, req);
+  uv__getaddrinfo_common_done(req, status);
+}
+
+static void uv__getaddrinfo_async_work(uv_loop_t* loop, uv__io_t* w, unsigned int fflags) {
+  READ_MACH_MSG_AND_HANDLE_DATA(w,
```
Comment (barbieri) [inline src/unix/getaddrinfo.c:62]:
```c
uv_getaddrinfo_t* req = container_of(w, uv_getaddrinfo_t, async_addrinfo_io);
```

should allow you to remove `w->data` asked by @bnoordhuis

File: src/unix/getaddrinfo.c [hunk 11]
```diff
@@ -135,6 +158,24 @@ static void uv__getaddrinfo_done(struct uv__work* w, int status) {
 }
 
 
+static void uv__getaddrinfo_work(struct uv__work* w) {
+  uv_getaddrinfo_t* req;
+  int err;
+
+  req = container_of(w, uv_getaddrinfo_t, work_req);
```
Comment (barbieri) [inline src/unix/getaddrinfo.c:165]: @cabelitos like this one (but different member, since this is the `uv__work` here, there it's the `uv__io_t` there)

File: include/uv/darwin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bnoordhuis) [reply include/uv/darwin.h:40]: It's applies to all three, they all change the ABI (the size of the structs they're used in.)

File: src/unix/getnameinfo.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnoordhuis) [reply src/unix/getnameinfo.c:173]: How so? `getnameinfo_async_start()` does some initialization work that can fail but there's a decent chance regular getnameinfo() will still work.

[the remark this answers is not in the document]
Comment (barbieri) [reply src/unix/getnameinfo.c:173]: this is very unlikely.

OTOH handling both cases makes the code more complex, for instance `uv_cancel()` (ok, that one is like 2 lines of code more), but I'd rather go with a reduce code path and see if that comes back in practice (which I think will never come back).

File: include/uv/darwin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply include/uv/darwin.h:40]: ouch, then it becomes a major issue... when is it ok to force the users to recompile (that would be enough)?

another approach (but then it makes the handle of "fallback" even harder) is to use the `struct uv__work work_req` space (using an `union` would make it more future proof):

```c
struct uv__io_s {
  uv__io_cb cb; // ptr
  void* pending_queue[2]; // 2 ptr
  void* watcher_queue[2]; // 2 ptr
  unsigned int pevents; // 4 bytes
  unsigned int events;  // 4 bytes
  int fd; // 4 bytes
}; //  5 ptr + 12 bytes = 32 or 52 bytes

struct uv__work {
  void (*work)(struct uv__work *w); // ptr
  void (*done)(struct uv__work *w, int status); // ptr
  struct uv_loop_s* loop; // ptr
  void* wq[2]; // 2 ptr
}; // total: 5 ptr = 20 or 40 bytes

// since sizeof(uv__io_s) > sizeof(uv_work), we must do dynamic allocation:

struct uv_getaddrinfo_libinfo_io {
  void *_zero; // uv__work.work == NULL could be used to check what's in use
  uv__io_t *io; // must be allocated as it won't fit into struct uv__work
  void *io_data; // will be the container since we cannot use container_of() anymore :-(
}; // total: 3 ptr = 12 or 24 bytes


#define UV_GETADDRINFO_PRIVATE_FIELDS                                         \
  union { \
     struct uv__work work_req; \
     struct uv_getaddrinfo_libinfo_io io_req; \
  } \
// just to illustrate, would need macros to be properly defined, only on macOS with the libinfo enabled
```

then if `work_req.work == NULL` we're using it as an async load and we must handle it as `io_req`. Since `uv__io_t` is too big, we need to dynamically allocate it... but there is no need to keep an external data structure to point back to it (hashes or the likes).

Ugly, but should work if we can't ask people to recompile.

File: src/unix/getaddrinfo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (cabelitos) [reply src/unix/getaddrinfo.c:45]: I don't think we'll need to comment it, since we'll do a dlsym in here.

File: src/unix/kqueue.c [hunk 12]
```diff
@@ -319,6 +335,24 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
...
+          struct kevent events[1];
+          EV_SET(events + 0, fd, ev->filter, EV_DELETE, 0, 0, 0);
+          if (kevent(loop->backend_fd, events, 1, NULL, 0, NULL))
+            if (errno != ENOENT)
+              abort();
+        }
+        if ((ev->flags & EV_EOF) && (w->pevents & UV__POLLRDHUP))
+          revents |= UV__POLLRDHUP;
+      }
+#endif
```
[the remark this answers is not in the document]
Comment (cabelitos) [reply src/unix/kqueue.c:354]: I'll happily create a follow-up PR once this one is merged.

File: src/unix/getaddrinfo.c (hunk 10 above)
Comment (cabelitos) [reply src/unix/getaddrinfo.c:62]: I like the idea, however I think that @bnoordhuis asked to remove the async_addrinfo_io all together, right?

File: include/uv/darwin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (cabelitos) [reply include/uv/darwin.h:40]: I guess he meant to remove all additions and only use `loop->internal_fields`. However I have a question for you @bnoordhuis I was looking at the this field and it is already populated right here https://github.com/libuv/libuv/blob/4296fec7f50145e2a307f3db7ae22984713976a7/src/unix/loop.c#L40-L43

Did you mean to add a hash table in here  (at struct uv__loop_internal_fields_s) ?
https://github.com/libuv/libuv/blob/5c85d67bc781430f2db8f446795494aa8d399de8/src/uv-common.h#L368-L370

**Comment by cabelitos** [comment]
Thank @barbieri and @bnoordhuis for the review. I fixed all comments except the ones regarding the hash map, which I'm waiting your answers so we can discuss and proceed with the implementation.

File: test/test-threadpool-cancel.c (hunk 1 above)
Comment (cabelitos) [reply test/test-threadpool-cancel.c:194]: Yeah, you got my idea in here. I'll leave with the timeout set to 0.

File: src/unix/getnameinfo.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (cabelitos) [reply src/unix/getnameinfo.c:173]: I guess that we can't escape from some kind of checking in `uv_cancel()`, since we're switching to `dlopen`/`dlsym` and we'll only know if we are using the async version at runtime. However I think that in this section, we should just fail in case the async version returns an error code.

File: CMakeLists.txt (hunk 3 above)
Comment (bnoordhuis) [reply CMakeLists.txt:36]: To clarify: using dlopen() is a requirement.

File: include/uv/darwin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (bnoordhuis) [reply include/uv/darwin.h:40]: > Did you mean to add a hash table in here (at struct uv__loop_internal_fields_s) ?

That's right.

> when is it ok to force the users to recompile

When libuv 2.0 is released. :-)

File: src/unix/getnameinfo.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (bnoordhuis) [reply src/unix/getnameinfo.c:173]: I would appreciate it if you and @barbieri motivate why you feel that way. My perspective is that falling through to regular getnameinfo() adds very little complexity.

File: src/unix/getaddrinfo.c (hunk 10 above)
Comment (bnoordhuis) [reply src/unix/getaddrinfo.c:62]: That's right.

File: CMakeLists.txt (hunk 3 above)
Comment (barbieri) [reply CMakeLists.txt:36]: sure, with misusage I meant trying to build this on non-mac systems... but on macOS would be nice to have that enabled by default (thus checking helps there)

File: include/uv/darwin.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply include/uv/darwin.h:40]: @bnoordhuis is it okay to go with my suggested approach? It removes a lot of work to manage a hashtable and doesn't break the ABI. We can add the union or simply keep it internal with `#if sizeof(x) < sizeof(y)` checks to compile-time check there is no misusage.

TL;DR is `uv__work` can fit the pointers to a new memory block (which would be needed in a hash table as well, but no hashing/lookup). To differentiate from a regular `uv__work` and this new memory block container we'd use the first pointer (`uv__work->work`), if it's null it's used by the async/libinfo work and the other member will point to the newly allocated memory. As it's not expected to have `uv__work->work == NULL`, then it should be enough and not break anything else. Alignments and all are also preserved.

File: src/unix/getaddrinfo.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply src/unix/getaddrinfo.c:45]: but would be nice to know "where the heck these signatures came from?" :-)

File: src/unix/getaddrinfo.c (hunk 10 above)
Comment (barbieri) [reply src/unix/getaddrinfo.c:62]: @cabelitos by removing the `async_addrinfo_io` and reusing the `uv__work` storage you can add the pointer in there as well (see what I called `io_data`).

File: src/unix/getnameinfo.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (barbieri) [reply src/unix/getnameinfo.c:173]: reducing code paths helps to avoid bugs due lack of testing of such code paths... but ok, agreed it's not a major thing, let's fallback :-)

**Comment by cabelitos** [comment]
Hello @bnoordhuis sorry for the delay. I implemented everything that you requested (including the hash alternative). Could you please take a look when you have some time?

* It uses dlopen/dlsysm
* It uses a hash table to store extra data
* It fallbacks to the existing logic, in case the async goes wrong.

Thank you very much.

**Comment by cabelitos** [comment]
@bnoordhuis Fixed the problems related to ASAN. We're good to go! :]

File: src/unix/darwin.c [hunk 13]
```diff
@@ -38,12 +38,21 @@
 static uv_once_t once = UV_ONCE_INIT;
 static uint64_t (*time_func)(void);
 static mach_timebase_info_data_t timebase;
+#if defined(_USE_LIBINFO)
+static uv_once_t once_getaddrinfo = UV_ONCE_INIT;
+static uv_once_t once_getnameinfo = UV_ONCE_INIT;
```
Comment (barbieri) [inline src/unix/darwin.c:43]: since its a single flag, use a single init function that calls both.

**Comment by barbieri** [comment]
@cabelitos looks ok, but I really don't like the idea of the hash table, but since @bnoordhuis asked for it, let's see what he will say (I'm not a libuv approver, so ultimately it's his call).

The hash table adds lots of code and still is a very simplified version that is only to be used with pointers (the immediate use, so ok).

Not just that, but lookup is much slower than direct member access, memory locality is also worse (since it will allocate book-keeping structures) and there is still a need to allocate the structure (so that is not saved).

My suggestion would still to use the available memory storage in the existing structure, keep ABI stable until a new major release is done, then we can even save the extra memory allocation.

**Comment by bnoordhuis** [comment]
@cabelitos it looks like it's still a build-time switch? Why is that?

My apologies for not reviewing in depth but to review a bigger change I need a longer stretch of time than I currently have available.

**Comment by bnoordhuis** [comment]
This PR has bitrotted, unfortunately. I'm going to close it but if someone wants to pick it up again, please do.
